cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go

```diff
@@ -0,0 +1,381 @@
+// Contains tells whether a contains x.
+func Contains(a []string, x string) bool {
+	for _, n := range a {
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+// createPacketManagerRest sets up the client and returns
+// an packetManagerRest.
+func createPacketManagerRest(configReader io.Reader, discoverOpts cloudprovider.NodeGroupDiscoveryOptions, opts config.AutoscalingOptions) (*packetManagerRest, error) {
+	var cfg ConfigFile
+	if configReader != nil {
+		if err := gcfg.ReadInto(&cfg, configReader); err != nil {
+			klog.Errorf("Couldn't read config: %v", err)
+			return nil, err
+		}
+	}
+
+	if opts.ClusterName == "" && cfg.Global.ClusterName == "" {
+		klog.Fatalf("The cluster-name parameter must be set")
+	} else if opts.ClusterName != "" && cfg.Global.ClusterName == "" {
+		cfg.Global.ClusterName = opts.ClusterName
+	}
+
+	manager := packetManagerRest{
+		clusterName:       cfg.Global.ClusterName,
+		projectID:         cfg.Global.ProjectID,
+		apiServerEndpoint: cfg.Global.APIServerEndpoint,
+		facility:          cfg.Global.Facility,
+		plan:              cfg.Global.Plan,
+		os:                cfg.Global.OS,
+		billing:           cfg.Global.Billing,
+		cloudinit:         cfg.Global.CloudInit,
+		reservation:       cfg.Global.Reservation,
+		hostnamePattern:   cfg.Global.HostnamePattern,
+	}
+	return &manager, nil
+}
+
+func (mgr *packetManagerRest) listPacketDevices() *Devices {
+	var jsonStr = []byte(``)
+	packetAuthToken := os.Getenv("PACKET_AUTH_TOKEN")
+	url := "https://api.packet.net/projects/" + mgr.projectID + "/devices"
+	req, err := http.NewRequest("GET", url, bytes.NewBuffer(jsonStr))
+	req.Header.Set("X-Auth-Token", packetAuthToken)
+	req.Header.Set("Content-Type", "application/json")
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
+		// klog.Fatalf("Error listing nodes: %v", err)
+	}
+	defer resp.Body.Close()
+
+	fmt.Println("response Status:", resp.Status)
+	body, _ := ioutil.ReadAll(resp.Body)
+	var devices Devices
+	json.Unmarshal([]byte(body), &devices)
+	return &devices
+}
+
+// nodeGroupSize gets the current size of the nodegroup as reported by packet tags.
+func (mgr *packetManagerRest) nodeGroupSize(nodegroup string) (int, error) {
+	devices := mgr.listPacketDevices()
+	// Get the count of devices tagged as nodegroup members
+	count := 0
+	for _, d := range devices.Devices {
+		if Contains(d.Tags, "k8s-cluster-"+mgr.clusterName) && Contains(d.Tags, "k8s-nodepool-"+nodegroup) {
+			count++
+		}
+	}
+	fmt.Println(len(devices.Devices), count)
+	return count, nil
+}
+
+func randString8() string {
+	n := 8
+	rand.Seed(time.Now().UnixNano())
+	letterRunes := []rune("acdefghijklmnopqrstuvwxyz")
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(b)
+}
+
+// createNodes provisions new nodes on packet and bootstraps them in the cluster.
+func (mgr *packetManagerRest) createNodes(nodegroup string, nodes int) error {
```

Comment:
Isn't `nodes` ignored in this function? Meaning it will only create 1 node whatever the desired new nodes are?

Reply:
Good catch & great patch! Thank you @iaguis!